### PR TITLE
Use IMAP uid to retrieve new mails instead of "NOT SEEN" flag

### PR DESCRIPTION
**Pull request by prehor**
This PR introduces the following configuration options:
* `uid_tracking` - set to `true` to use IMAP uid instead of "NOT SEEN" flag. defaults to `false`
* `sincedb_path` - path to file with last processed IMAP uid. Defaults to `<path.data>/plugins/inputs/imap`

IMAP uid is always stored in the file `sincedb_path` regardles of the `uid_tracking` setting, so you can switch between "NOT SEEN" and "UID" tracking. In transition from the previous plugin version, first process at least one mail with `uid_tracking` set to `false` to save the last processed IMAP uid and then switch `uid_tracking` to `true`.

Fixies #31

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,4 +25,3 @@
  - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully,
    instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
  - Dependency on logstash-core update to 2.0
-
```
Comment (robbavey) [inline CHANGELOG.md]: Unnecessary change

File: lib/logstash/inputs/imap.rb [hunk 2]
```diff
@@ -29,11 +29,17 @@ class LogStash::Inputs::IMAP < LogStash::Inputs::Base
...
   config :expunge, :validate => :boolean, :default => false
   config :strip_attachments, :validate => :boolean, :default => false
-  
+
   # For multipart messages, use the first part that has this
   # content-type as the event message.
   config :content_type, :validate => :string, :default => "text/plain"
 
+  # Whether to use IMAP uid to track last processed message
+  config :uid_tracking, :validate => :boolean, :default => true
```
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: My fault. The value should be false, I used true for debugging and forgot to change it back. I will fix it.

File: lib/logstash/inputs/imap.rb [hunk 3]
```diff
@@ -87,16 +117,28 @@ def check_mail(queue)
...
+      @logger.info("Saving \"uid_last_value\": \"#{@uid_last_value}\"")
+
+      # Always save @uid_last_value so when tracking is switched from
+      # "NOT SEEN" to "UID" we will continue from first unprocessed message
+      File.write(@sincedb_path, @uid_last_value) unless @uid_last_value.nil?
       imap.store(id_set, '+FLAGS', @delete ? :Deleted : :Seen)
-  
+
+      # Stop mail fetching if it is requested
+      break if @stop_called.true?
```
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: Yes. I don't know the entire plugin API, sometimes I use some methods non-optimally. I will fix it.

File: lib/logstash/inputs/imap.rb [hunk 4]
```diff
@@ -87,16 +117,28 @@ def check_mail(queue)
         else
           queue << parse_mail(mail)
         end
+        @uid_last_value = item.attr["UID"]
+
+        # Stop mail fetching if it is requested
+        break if @stop_called.true?
```
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: I will fix it.

File: CHANGELOG.md (hunk 1 above)
Comment (prehor) [reply CHANGELOG.md]: Sorry. My editor do this automatically. Should I revert it?

Comment (robbavey) [reply CHANGELOG.md]: Yes please

File: lib/logstash/inputs/imap.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: Fixed in 312b48a

File: lib/logstash/inputs/imap.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: Fixed in b3339bf

File: lib/logstash/inputs/imap.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (prehor) [reply lib/logstash/inputs/imap.rb]: Fixed in b3339bf

File: CHANGELOG.md (hunk 1 above)
Comment (prehor) [reply CHANGELOG.md]: Fixed in 4248975

File: docs/index.asciidoc [hunk 5]
```diff
@@ -131,31 +133,54 @@ content-type as the event message.
...
   * Default value is `false`
 
 
 
+[id="plugins-{type}s-{plugin}-uid_tracking"]
+===== `uid_tracking`
+
+  * Value type is <<boolean,boolean>>
+  * Default value is `false`
+
```
Comment (robbavey) [inline docs/index.asciidoc]: Can we have a small explanation as to what `uid_tracking` is, and a description of the differences in behaviour  between having this flag set and not?

Comment (prehor) [reply docs/index.asciidoc]: Fixed in e0a12bb

File: docs/index.asciidoc [hunk 6]
```diff
@@ -173,11 +173,17 @@ NOTE: it must be a file path and not a directory path.
...
+the UID of the last processed mail is not yet known, the unread mails are
+first downloaded and the UID of the last processed mail is saved. From
+this point on, if `uid_tracking` is set to `true`, all new mail will be
+downloaded regardless of whether they are marked as read or unread. This
+allows users or other services to use the mailbox simultaneously with the
+IMAP input plugin. UID of the last processed mail is always saved regardles
+of the `uid_tracking` value, so you can switch its value as needed. In
+transition from the previous IMAP input plugin version, first process at least
+one mail with `uid_tracking` set to `false` to save the UID of the last
+processed mail and then switch `uid_tracking` to `true`.
```
Comment (robbavey) [inline docs/index.asciidoc]: Thank you!

**Comment by prehor** [comment]
Wait a minute please, a little improvement is on the way.

**Comment by prehor** [comment]
Hi @robbavey, I'm done. Could you please review the changes?

**Comment by robbavey** [comment]
@prehor Let me know if you are good with these changes being merged, and I will go ahead and merge and publish the gem. 

Thanks again for the contribution!

**Comment by prehor** [comment]
I think that it is finished. Today I downloaded approximately 10.000 emails from Gmail and plugin was working well.

**Comment by prehor** [comment]
Hi @robbavey . Thanks for your help. Could you please spend some time on my second PR # 35? I just rebase it. Thanks.

**Comment by robbavey** [comment]
@prehor Thanks for the pointer - I've just assigned it to myself to review and will take a look at it shortly. Thanks again!
